Replace the global config dict with new class ManimConfig

leotrs opened the pull request:
## Motivation
Short version: the config system is Bad. This PR makes it Less Bad by introducing a new dict-like class `ManimConfig`. Instances of this class behave pretty much like the previous global `config` dict, but maintain internal consistency, among other features. Long version: see #337. 

## List of Changes
The diff is long. Here I explain the changes file by file. Hopefully this will make reviewing easier. Note most of the changes are due to the removal of `file_writer_config`.
- Take the logger logic from the file `manim/config/utils.py` into a new file `manim/config/logger.py`. This also explains the diff in `tests/control_data/logs_data/BasicSceneLoggingTest.txt`.
- This PR removes the global dicts `file_writer_config`, `camera_config` and `config`, and replaces them with a (still global) instance of the `ManimConfig` class. This global is also called `config`. This accounts for the diff in `manim/__main__.py`, `manim/camera/camera.py`, `manim/config/__init__.py`, `manim/grpc/impl/frame_server_impl.py `, `manim/mobject/mobject.py`, `manim/mobject/svg/text_mobject.py`, `manim/utils/module_ops.py`, `tests/utils/GraphicalUnitTester.py`, `tests/test_copy.py`, `tests/test_container.py`, `tests/helpers/graphical_units.py`, and `manim/utils/tex_file_writing.py`.
- Add the config options `video_dir`, `tex_dir`, `text_dir`, `images_dir` to `manim/config/default.cfg`. These DO NOT use the placeholder syntax used by `configparser`. Instead, they use f-string syntax, as used by `ManimConfig._get_dir()`.
- The file `manim/config/main_utils.py` has two modifications: the `init_dirs` function has been moved to `manim/config/utils.py` and the function `update_config_with_cli` has disappeared and its logic moved inside `ManimConfig.digest_args()`.
- `manim/constants.py`: we have changed the default resolution from "production" to "high" (I believe either @kolibril13 or @huguesdevimeux did this) so I had to update one line in this file.
- added a test to test the -s flag to `tests/test_scene_rendering/test_cli_flags.py`
- remove the file `tests/test_scene_rendering/test_caching_relalted.py`, which had the same content as `tests/test_scene_rendering/test_caching_related.py` (note the typo). This was already removed before, but I don't know how it snuck into my branch :shrug: 
- The files `tests/test_logging/test_logging.py`, `tests/test_scene_rendering/test_caching_related.py`, `tests/test_scene_rendering/conftest.py`, and `tests/test_cli_flags.py` each have a one-line quality of live improvement. 
- `tests/test_config.py` has been modified because `ManimConfig` no longer allows adding/updating keys that did not previously exist.
- The diff in `manim/scene/scene_file_writer.py` is explained by the removal of `scene_file_writer` EXCEPT for the new logic in lilne 128. This change in logic now checks whether a file exists before using it (which makes sense and has nothing to do with the `ManimConfig` class!)
- The diff in `manim/scene/scene.py` is explained by the removal of `scene_file_writer` EXCEPT for the change in logic in lines 100-102 (some child objects are now made to use the global config objects instead of passing around things like `**camera_config`) and the logic in line 128, which now correctly restores the previous state of `skip_animations` (before it was always just set to False, which makes no sense).
- `manim/config/utils.py` - this is the big one. The `init_dirs()` function comes from `manim/config/main_utils.py`, and the parser logic has moved to `manim/config/parser.py`. The big change: it now contains the definition of the two main classes `ManimConfig` and `ManimFrame`. More on these later.


## Explanation for Changes
Ok the main idea of `ManimConfig` is the following. It behaves like a dict but it keeps track of all the logic between different config options. To behave like a dict, it implements some dunder methods such as `__iter__`, `__len__`, `__contains__`, `__getitem__`, `__setitem__`. However, for documentation purposes, each config option is *also an attribute*. For example the config option `write_to_movie` can be accessed as `config["write_to_movie"]` or as `config.write_to_movie`. The former is backwards compatible, while the latter allows for easy documentation and auto-completion. To implement this behavior, there is a private attribute `self._d` which is an actual dictionary which contains the config options as keys. The user *should never* access it directly. Most of the get/set code touches `self._d` in one way or another. Furthermore, `ManimConfig` does not allow to add new keys or remove existing keys (see `__delitem__` and `__delattr__`). 

`ManimConfig` can only be initialized by passing a `ConfigParser` instance. This will usually contain the options set in `manim/config/default.cfg`, as well as any other config files such as the folder-wide or the user-wide. The global `config` is instantiated on line 25 of `manim/config/__init__.py` with the config parser object returned by `manim/config/utils.py::make_config_parser`. During initialization, the function `digest_parser()` is called, which consumes the parser. After this, all config files have been parsed and digested into the global `config`, and manim is ready to go.

Now, if manim is being called from the command line, then `manim/__main__.py` will call `config.digest_args(args)`. This will take the parsed `sys.argv` args and update the necessary config options. Note that the options from the config files and from the CLI flags are parsed at different times. Because of this, there is no longer any need of hacky things such as trying to guess whether manim was being used from the command line or not. This is a huge gain in readability I think.

I have left placeholder methods `ManimConfig.digest_dict` and `ManimConfig.digest_file`. The latter will eventually allow to read a configuration file on the fly. This feature was requested by @ManimCommunity/tests and will simplify testing a lot.

Finally, the rest of the code of `ManimConfig` is the definition of a ton of properties. Each config option is now a property. In this way, we have complete control of what happens when a config option is read or set. In this way, we keep internal consistency of all options. For example, see `ManimConfig.tex_template_file` or `ManimConfig.dry_run`. Look also at `ManimConfig.quality` which now allows to change the width, height, and frame rate with a single line of code `config["quality"] = "low_quality"` (@eulertour recently requested a `set_quality` method :wink: ). The rest of the properties are mostly boilerplate. I chose to use the syntax `config_opt = property(...)` because it reduces boilerplate and wasted space (though I am not married to this). There are a few helpers that help with this boilerplate (`_set_from_list`, `_set_boolean`, `_set_str`, `_set_between`, `_set_pos_number`).

The `ManimConfig` class is meant to contain *all* the configuration in one place, a single source of truth, and it replaces `file_writer_config`, and most of `config`, and is meant to be used **mostly internally**. HOWEVER, I have also added a new class, `ManimFrame`, which is meant to contain the config options that are related to frame/video geometry and are meant to be used **mostly by the user**. The global frame is called `frame` and is instantiated in line 26 of `manim/config/__init__.py` and depends intimately on the global `ManimConfig` object. In short, `ManimFrame` just reflects a bunch of config options set in `ManimConfig` (see bottom of the file `manim/config/utils.py`) but a `ManimFrame` is completely immutable. My intention is to allow the user to do something like `from manim import frame as F` and then do something like `F.pixel_width` or `F.top`. (This is something @kolibril13 has been asking for for a while!) However, I'm not married to the current implementation of `ManimFrame` and that is subject to change. **Please let's focus right now on `ManimConfig` first!**

That's all folks. I've done my best to explain the inner workings of this beast. Please let me know if I missed anything.


## Testing Status
Passes locally. Added a new test for the -s flag.


## Further Comments
Right now the `ManimFrame` class is implemented but never used. I wanted to reduce the size of this PR as much as possible before actually using it. If you have general comments on it, please leave them here. If you have implementation comments, let's discuss them in the next PR.

Once the review of this stabilizes a bit, I will write the changelog right before merging. I will add extensive documentation in a separate PR.

I still want to add a test for `--disable_caching` before merging this, but haven't figured out how. Help please @ManimCommunity/tests.


## Acknowledgement
- [x] I have read the [Contributing Guidelines](https://github.com/ManimCommunity/manim/wiki/Documentation-guidelines-(WIP))

eulertour commented:
Everything here looks good to me, it's just a matter of addressing the problems with the merge and docs.

leotrs commented:
Just pushed a new commit that fixes the -s flag when the scene contains no animations. Thanks @kolibril13 for reporting. (It also removes a couple of repeated keys.)

@ManimCommunity/tests could you please double-check the test that was added in[https://github.com/ManimCommunity/manim/pull/620/commits/e74289e57955cf160239a6337370f7a8ef4a2d54](url)? I'm not sure if there's a better way of doing this.

huguesdevimeux commented on manim/config/utils.py:
```diff
@@ -70,201 +61,825 @@ def make_config_parser():
+            self.save_pngs = False
+            self.save_as_gif = False
+        else:
+            raise ValueError(
+                "It is unclear what it means to set dry_run to "
+                "False.  Instead, try setting each option "
+                "individually. (write_to_movie, write_alll, "
+                "save_last_frame, save_pngs, or save_as_gif)"
```
**Comment** This sounds really fair. But I think it would be nice if dry_run is set back to false, then it revert to original state of `ManimConfig` object.

(see the post of the review for the revert thing)

leotrs replied:
I thought of this, but the problem is *what is the "original state" of ManimConfig*? Is it the default state? Is it the state after you read some .cfg files? Is it the state just before you set dry_run to True?

huguesdevimeux commented on manim/scene/scene_file_writer.py:
```diff
@@ -386,11 +387,11 @@ def finish(self):
+            if config["flush_cache"]:
                 self.flush_cache_directory()
             else:
                 self.clean_cache()
```
**Question** This used to be there
```py
        if config["save_last_frame"]:
            self.scene.update_frame(ignore_skipping=True)
```
What happened? Where/why did you move it ? 
I saw on a precedent commit that you changed `ignore_skipping` to False, which is IMO not a good idea, but I can't even see theses lines on HEAD lol

leotrs replied:
This was moved to `Renderer.finish()`, by a previous PR, not this one.

huguesdevimeux reviewed (approved):
Amazing. 

Just a request, that is not implemented (or didn't I noticed) : 
A method to revert manim config to the previous state. 
I don't know if iy would be particularly useful, but who knows. 


And ... I know you're going to hate me for that but ...


This is typically the type of thing that needs ton of tests, if we don't want it to well, break at every PR. 
_Could you add some please?_

leotrs commented:
> Just a request, that is not implemented (or didn't I noticed) :
> A method to revert manim config to the previous state.
> I don't know if iy would be particularly useful, but who knows.

`tempconfig` exists! It even has a couple of tests already.

> This is typically the type of thing that needs ton of tests, if we don't want it to well, break at every PR.
> _Could you add some please?_

Yeah, I want to add a few tests for some more CLI flags, at least. Do you have suggestions on what else I should be testing?

leotrs commented on tests/test_scene_rendering/test_cli_flags.py:
```diff
@@ -61,3 +62,51 @@ def test_n_flag(tmp_path, simple_scenes_path):
+    assert is_empty, "running manim with -s flag rendered a video"
+
+    is_empty = not any((tmp_path / "images" / "simple_scenes").iterdir())
+    assert not is_empty, "running manim with -s flag did not render an image"
```
I love when you can read python code almost as if it were English.

leotrs commented on manim/scene/scene_file_writer.py:
```diff
@@ -126,6 +125,8 @@ def add_partial_movie_file(self, hash_animation):
         hash_animation : str
             Hash of the animation.
         """
+        if not hasattr(self, "partial_movie_directory"):
+            return
```
Well all of the rendering logic is super broken. But the problem in this particular instance is that `partial_movie_directory` is only set when `config["write_to_movie"]` is set to True. So during refactoring, this function was being called sometimes even when `config["write_to_movie"]` was set to False. Now, even if it is called in this case, it doesn't error out, it simply doesn't do anything. Not the cleanest, but hey.

leotrs commented on manim/constants.py:
```diff
@@ -148,5 +148,5 @@ class MyText(Text):
 }
 
-DEFAULT_QUALITY = "production_quality"
+DEFAULT_QUALITY = "high_quality"
```
I thought this had been changed in a different PR?

leotrs commented:
Update: tests pass locally and docs build.

However, there are a few tests that break only in python3.6. I'll investigate.

leotrs commented:
All tests pass! Now I'll add some more tests as requested by @huguesdevimeux

leotrs commented:
(Hopefully) last update: tests pass, docs build. Added a few more tests and cleaned up some unnecessary code.

@eulertour I believe there are no tests for the js renderer so if you want to test that out before merging, that'd be great.

Otherwise, I move we merge this ASAP because rebasing is killing me. I will write documentation in a separate PR.

behackl commented on manim/config/utils.py:
```diff
@@ -56,215 +47,885 @@ def config_file_paths():
-def make_config_parser():
-    """Make a ConfigParser object and load the .cfg files."""
+def make_config_parser(custom_file=None):
+    """Make a ConfigParser object and load the .cfg files.
```
Detailed documentation for module-level functions is currently not rendered it seems (the oneline-summary is rendered in the list of functions, but documentation for parameters and notes etc. are not) -- we should take care of this in a separate PR, though.

behackl reviewed (approved):
This looks great; amazing job! I love how readable especially arg parsing suddenly is. :rocket: 

I only have two issues/comments, but they can both be settled in follow-up PRs.

leotrs commented on docs/source/reference.rst:
```diff
@@ -128,6 +128,6 @@ Other modules
    :toctree: reference
 
-   _config
+   config
```
Agreed this is annoying. I think the way to go is to rename the folder and subpackage from `manim/config/` to `manim/_config/`. The problem before is that its name was `config` (just like the global dict itself), and so sphinx was having trouble differentiating both. So I tried importing it as `_config`. But that didn't work. I think I'll just change its name to `manim/_config/` or even `manim/config_pkg/`. Same with the logger.

behackl replied:
I agree, renaming `config` (the directory) probably makes sense. I suggest `configuration`, but I don't mind other options either. :-)

leotrs commented:
@eulertour do you want to test JS renderer stuff before I merge and inevitably break everything?

eulertour commented:
I'm actually fine with you merging it first, since after making the Cairo renderer the J's renderer will have to be reworked anyway
